pilot/pkg/xds/endpoint_builder.go

```diff
@@ -293,12 +294,19 @@ func (b *EndpointBuilder) buildLocalityLbEndpointsFromShards(
 	// The shards are updated independently, now need to filter and merge for this cluster
 	for _, clusterID := range keys {
 		endpoints := shards.Shards[clusterID]
+
 		// If the downstream service is configured as cluster-local, only include endpoints that
 		// reside in the same cluster.
 		if isClusterLocal && (cluster.ID(clusterID) != b.clusterID) {
 			continue
 		}
 		for _, ep := range endpoints {
+			// cross-cluster WorkloadEntry may come from a shard with a different cluster ID
+			// AugmentLabels adds an internal label to indicate individual endpoint origin
+			if epCluster := ep.Labels[label.TopologyCluster.Name]; epCluster != "" && epCluster != b.clusterID {
```

Comment:
Should we just set the cluster field on the `IstioEndpoint`? Seems like the existing cluster-local logic would just work ... or am I missing something?

Reply:
The above cluster-local logic is purely based off the shard names and doesn't look at istio endpoints. I don't think there is a cluster field, just this label.

Reply:
I might try to push this into the DiscoverabilityPolicy so we don't have to dupe this logic across here and https://github.com/istio/istio/blob/462811d9d1503fdc6069b81f1812276b68dff2e7/pilot/pkg/networking/core/v1alpha3/cluster_builder.go#L502

Reply:
Ah I see ..  so the issue is that `b` can contain endpoints from multiple clusters? Then yes, I think we should remove the cluster-local check outside of the loop and instead have the check on a per-endpoint basis.

I had considered adding this to `DiscoverabilityPolicy`, but the issue is that the policy would have to do the check to see if the service is cluster-local each time, which is a bit costly.

Reply:
`b.clusterId` is the proxy's cluster

The original issue was that ` shards.Shards` had all WorkloadEntry endpoints in a shard `""` which would never be usable in cluster-local mode. 

Now we set a cluster ID on that registry, but a new issue is that shard names need more info to separate k8s cluster-1 and serviceentry cluster-1; shard name can't be considered a safe cluster id..

Even when that's solved, we update the shards in cluster-1 containing a service with an endpoint from cluster-2's workloadentry (see AppendWorkloadHandler calls).

tl;dr, shards become an invalid way to determine which cluster an IstioEndpoint is a part of

Reply:
> Now we set a cluster ID on that registry, but a new issue is that shard names need more info to separate k8s cluster-1 and serviceentry cluster-1; shard name can't be considered a safe cluster id..

Ah got it ... so there are multiple service registries that attempt to update the same shard but with different endpoints, causing collisions. Makes sense. We should definitely break any reliance on shard being a cluster ID.

> Even when that's solved, we update the shards in cluster-1 containing a service with an endpoint from cluster-2's workloadentry (see AppendWorkloadHandler calls).

I'm not quite sure I understand what you mean here. Can you clarify?

Reply:
I thought we could have a Service in cluster-a select a WorkloadEntry in cluster b

I don't think it can actually happen:

https://github.com/istio/istio/blob/b03691f9e0cc1d8f9560ab72ad75cbe326ff5f36/pilot/pkg/serviceregistry/kube/controller/multicluster.go#L204-L205

But our FakeDiscoveryServer in tests does that

https://github.com/istio/istio/blob/e24ae0a21c892c604195f8bc7125a7cc2746f568/pilot/pkg/xds/fake.go#L251-L258

Reply:
So in that case – rather than relying completely on the label, maybe we could have something like 

```
type shardKey struct {cluster cluster.ID provider provider.ID}
``

Reply:
> I thought we could have a Service in cluster-a select a WorkloadEntry in cluster b

Not sure I follow ... Doesn't Istio have a single merged definition of a given service, with endpoints spread across multiple clusters?